community/kernel/src/main/java/org/neo4j/kernel/impl/store/GeometryType.java

```diff
@@ -0,0 +1,164 @@
+    public abstract int calculateNumberOfBlocksUsedForGeometry( long firstBlock );
+
+    private static int getGeometryType( long firstBlock )
+    {
+        return (int) ((firstBlock & 0x00000000F0000000L) >> 28);
+    }
+
+    private static int getDimension( long firstBlock )
+    {
+        return (int) ((firstBlock & 0x0000000F00000000L) >> 32);
+    }
+
+    private static int getCRSTable( long firstBlock )
+    {
+        return (int) ((firstBlock & 0x000000F000000000L) >> 36);
+    }
+
+    private static int getCRSCode( long firstBlock )
+    {
+        return (int) ((firstBlock & 0x00FFFF0000000000L) >> 40);
+    }
+
+    private static boolean isFloatPrecision( long firstBlock )
+    {
+        return ((firstBlock & 0x0100000000000000L) >> 56) == 1;
+    }
+
+    public static int calculateNumberOfBlocksUsed( long firstBlock )
+    {
+        GeometryType geometryType = find( getGeometryType( firstBlock ) );
+        if ( geometryType == null )
+        {
+            return PropertyType.BLOCKS_USED_FOR_BAD_TYPE_OR_ENCODING;
+        }
+        else
+        {
+            return geometryType.calculateNumberOfBlocksUsedForGeometry( firstBlock );
+        }
+    }
+
+    private static GeometryType find( int gtype )
+    {
+        for ( GeometryType type : GeometryType.values() )
+        {
+            if ( type.gtype == gtype )
+            {
+                return type;
+            }
+        }
+        return null;
+    }
+
+    public static Value decode( PropertyBlock block )
+    {
+        return decode( block.getValueBlocks(), 0 );
+    }
+
+    public static Value decode( long[] valueBlocks, int offset )
+    {
+        long firstBlock = valueBlocks[offset];
+        int gtype = getGeometryType( firstBlock );
+        int dimension = getDimension( firstBlock );
+
+        if ( isFloatPrecision( firstBlock ) )
+        {
+            throw new UnsupportedOperationException( "Float precision is unsupported in Geometry properties" );
+        }
+        if ( dimension > 3 )
+        {
+            throw new UnsupportedOperationException( "Points with more than 3 dimensions are not supported in the PropertyStore" );
+        }
+        CoordinateReferenceSystem crs = CoordinateReferenceSystem.get( getCRSTable( firstBlock ), getCRSCode( firstBlock ) );
+        return find( gtype ).decode( crs, dimension, valueBlocks, offset );
+    }
+
+    public static long[] encodePoint( int keyId, CoordinateReferenceSystem crs, double[] coordinate ) throws IllegalArgumentException
+    {
+        if ( coordinate.length > 3 )
+        {
+            // One property block can only contains at most 4x8 byte parts, one for header and 3 for coordinates
+            throw new UnsupportedOperationException( "Points with more than 3 dimensions are not supported in the PropertyStore" );
+        }
+
+        long keyAndType = keyId | (((long) (PropertyType.GEOMETRY.intValue()) << StandardFormatSettings.PROPERTY_TOKEN_MAXIMUM_ID_BITS));
```

Comment:
please use some local variable for StandardFormatSettings.PROPERTY_TOKEN_MAXIMUM_ID_BITS takes a lot of space and please described desired format somewhere in java doc